### PR TITLE
Fix add_class_trait in the presence of subclasses

### PR DESCRIPTION
**Pull request by mdickinson**
This PR fixes the issue described in #1460, and adds some tests for `add_class_trait`. Prior to this PR, there were no tests at all for `add_class_trait`.

The PR also does a drive-by fix of a bug involving prefix traits and an invalid use of `sort`, leading to the error `TypeError: sort() takes no positional arguments`.

Unfortunately, one of the newly added tests is failing as part of the full test suite, but passing standalone. It appears that we have a test interaction - some test somewhere in the test suite is mutating the class traits on `HasTraits`. I'll try to track it down.

**Checklist**
- [x] Tests
- [ ] Update API reference (`docs/source/traits_api_reference`)  N/A
- [ ] Update User manual (`docs/source/traits_user_manual`)  N/A
- [ ] Update type annotation hints in `traits-stubs`  N/A

### CODE REVIEW

**Comment by mdickinson** [comment]
> I'll try to track it down.

The culprit is test_abc.py, in combination with issue #58: https://github.com/enthought/traits/blob/249d64813b633390616af4e57ceb19a553c250ef/traits/tests/test_abc.py#L33

**Comment by mdickinson** [comment]
The documentation build issue that's causing the CI to fail on this PR should be fixed in #1462.

File: traits/has_traits.py [hunk 1]
```diff
@@ -1135,14 +1137,39 @@ def add_class_trait(cls, name, *trait):
             trait = trait_for(trait[0])
 
         # Add the trait to the class:
-        cls._add_class_trait(name, trait, False)
+        cls._add_class_trait(name, trait, is_subclass=False)
```
Comment (mdickinson) [inline traits/has_traits.py:1140]: Drive-by code cleanup: it's not _required_ that the `is_subclass` parameter is passed by name, but it makes the code more readable.

File: traits/has_traits.py [hunk 2]
```diff
@@ -1161,7 +1188,7 @@ def _add_class_trait(cls, name, trait, is_subclass):
             prefix_list.append(name)
 
             # Resort the list from longest to shortest:
-            prefix_list.sort(lambda x, y: len(y) - len(x))
+            prefix_list.sort(key=lambda x: -len(x))
```
Comment (corranwebster) [inline traits/has_traits.py:1191]: Perhaps?
```suggestion
            prefix_list.sort(key=lambda x: len(x), reverse=True)
```

**Review by corranwebster — APPROVED** [review]
While orthogonal to main idea of the PR, there aren't tests for the `is_subclass` behaviour where there is an error as described in the new docstring for `_add_class_trait`.

Otherwise LGTM.

File: traits/has_traits.py (hunk 2 above)
Comment (mdickinson) [reply traits/has_traits.py:1191]: Yep, that works too, but I think I'd rather keep this consistent with the code here: https://github.com/enthought/traits/blob/d8d4e8ffb502e31375e5706b2b989ad166839571/traits/has_traits.py#L646

(We could change both, of course.)

**Comment by mdickinson** [comment]
> there aren't tests for the `is_subclass` behaviour where there is an error

Yes, that should be fixed. (Though the use of `TraitError` here bugs me: it feels like the wrong exception, and writing a test for it bakes in that wrongness a little bit harder. That's not a good excuse not to write that test, though.)

**Comment by mdickinson** [comment]
> writing a test for it bakes in that wrongness a little bit harder

Though I guess I've already done that by documenting it. :-(

**Comment by mdickinson** [comment]
> there aren't tests for the `is_subclass` behaviour where there is an error

Tests added.

File: traits/has_traits.py (hunk 2 above)
Comment (mdickinson) [reply traits/has_traits.py]: > (We could change both, of course.)

Done.
